templates/CHP.iesopt.template.yaml

```
@@ -0,0 +1,221 @@
+    capacity: <heat_max> out:heat
+  
+  aux_node:
+    type: Node
+    enabled: <mode> == extraction
+    carrier: electricity
+  
+  power:
+    type: Connection
+    enabled: <mode> == extraction
+    node_from: <self>.aux_node
+    node_to: <power_to>
+    lb: 0
+
+# TODO: discuss that these functions get really hard to read/maintain when they grow larger (since in a YAML)
+#       should we just default to addons?
+
+functions:
+  validate: |
+    # TODO: A lot of additional validation checks.
+
+    # Check mandatory parameters.
+    @check !isnothing(this.get("mode"))
+    @check !isnothing(this.get("fuel"))
+    @check !isnothing(this.get("fuel_from"))
+    @check !isnothing(this.get("power_to"))
+
+    # Check mode.
+    @check this.get("mode") in ["extraction", "condensing", "backpressure"]
+
+    # Check for non-negativity.
+    @check isnothing(this.get("power_max")) || (this.get("power_max") >= 0)
+    @check isnothing(this.get("heat_max")) || (this.get("heat_max") >= 0)
+    @check isnothing(this.get("power_at_heat_max")) || (this.get("power_at_heat_max") >= 0)
+    @check isnothing(this.get("power_ratio")) || (this.get("power_ratio") >= 0)
+    @check isnothing(this.get("power_loss_ratio")) || (this.get("power_loss_ratio") >= 0)
+    @check isnothing(this.get("efficiency_condensing")) || (this.get("efficiency_condensing") >= 0)
+    @check isnothing(this.get("efficiency_backpressure")) || (this.get("efficiency_backpressure") >= 0)
+    
+    # Mode dependent checks.
+    if this.get("mode") == "backpressure"
+        @check !isnothing(this.get("heat_to"))
+        @check !isnothing(this.get("efficiency_backpressure"))
+    elseif this.get("mode") == "condensing"
+        @check isnothing(this.get("heat_to"))
+        @check !isnothing(this.get("efficiency_condensing"))
+    elseif this.get("mode") == "extraction"
+        @check !isnothing(this.get("heat_to"))
+        @check !isnothing(this.get("efficiency_condensing")) || !isnothing(this.get("efficiency_backpressure"))
+    end
+
+  prepare: |
+    cm = this.get("power_ratio")
+    cv = this.get("power_loss_ratio")
+
+    fuel = this.get("fuel")
+    to_p = this.get("power_to")
+    to_co2 = this.get("co2_to")
+
+    if this.get("mode") == "backpressure"
+        to_h = this.get("heat_to")
+
+        # -------------------------------------------------
+        # Prepare: power_max, heat_max, power_ratio
+        # -------------------------------------------------
+        p_max = this.get("power_max")
+        h_max = this.get("heat_max")
+        cm = this.get("power_ratio")
+
+        if isnothing(p_max)
+            power_max = cm * h_max
+            this.set("power_max", power_max)
+        end
+        
+        if isnothing(h_max)
+            # Currently, nothing to do here (`h_max` is not actively used in backpressure mode).
+        end
+        
+        if isnothing(cm)
+            cm = p_max / h_max
+            this.set("power_ratio", cm)
+        end
```

Being super picky (or will you prevent the input-combination 'cm+h_max+p_at_h_max' in the validate section):

```suggestion
        p_at_h_max= this.get("power_at_heat_max")
        
        if isnothing(p_at_h_max)
            p_at_h_max=p_max
        end

        if isnothing(cm)
            cm = p_at_h_max / h_max
            this.set("power_ratio", cm)
        end
```

---

After discussion => we do not want want to allow over-specification, therefore we check for that in the validate, and then do not need to care about any over-specified combinations.